File: kalitech-engine/src/main/resources/kalitech/engine/modules/Camera/modes/third.js
// FILE: Scripts/player/modes/third.js
"use strict";

const U = require("../camUtil.js");

function clamp(v, lo, hi) {
    return v < lo ? lo : (v > hi ? hi : v);
}

function lerp(a, b, t) {
    return a + (b - a) * t;
}

function smoothstep(t) {
    t = clamp(t, 0, 1);
    return t * t * (3 - 2 * t);
}

function expAlpha(rate, dt) {
    return 1.0 - Math.exp(-Math.max(0, rate) * Math.max(0, dt));
}

class ThirdPersonCameraMode {
    constructor() {
        this.id = "third";
        this.meta = {supportsZoom: true, hasCollision: true, numRays: 8, playerModelVisible: true};

        // CP2077-style framing (over-the-shoulder with soft recenter)
        this.pivotOffset = {x: 0.0, y: 1.46, z: 0.0};

        this.shoulderX = 0.42;
        this.shoulderAimX = 0.26;
        this.verticalLift = 0.14;

        this.forwardLead = 0.28;
        this.forwardLeadAim = 0.08;
        this.leadSmooth = 12.0;

        this.pivotSmoothPos = 22.0;
        this.pivotSmoothY = 26.0;

        this.camSmoothPos = 18.0;
        this.camSmoothY = 20.0;

        this.recenterEnabled = true;
        this.recenterRate = 2.2;
        this.recenterDeadZone = 0.38;
        this.recenterMax = 1.25;

        this.pitchOrbitScale = 1.0;
        this.pitchMin = -1.15;
        this.pitchMax = 0.90;
        this.pitchSoft = 0.18;

        // Collision tuning (pear corridor)
        this.collisionEnabled = true;

        this.camRadius = 0.28;
        this.nearRadius = 0.06;
        this.pearK = 1.9;
        this.pearSamples = 9;

        this.surfacePadding = 0.08;
        this.obstaclePasses = 2;

        this.useTerrainHeight = true;
        this.terrainWorld = true;

        this.floorPadding = 0.22;
        this.slopePadScale = 0.45;

        this.groundRayLift = 1.25;
        this.maxRayLenDown = 12.0;
        this.groundSnapPen = 0.60;

        this.zoomRadiusBoost = 0.10;
        this.zoomNearBoost = 0.06;
        this.zoomFloorBoost = 0.20;
        this.zoomBoostStart = 6.0;
        this.zoomBoostFull = 26.0;

        this.debugCapsule = true;
        this.debugGroundCapsule = true;

        this._init = false;

        this._pivot = {x: 0, y: 0, z: 0};
        this._cam = {x: 0, y: 0, z: 0};
        this._lead = {x: 0, y: 0, z: 0};

        this._shoulderSide = +1;
    }

    _ensureModeConfig(ctx) {
        let mc = ctx.modeConfig;
        if (!mc) {
            mc = {};
            ctx.modeConfig = mc;
        }
        return mc;
    }

    _applyCollisionOverrides(ctx, dist) {
        // Publish collision tuning as modeConfig (baseline), zones will blend over it later.
        const mc = this._ensureModeConfig(ctx);

        // If a volume zone explicitly disables collision, respect it by not overriding.
        const zoRaw = ctx.zoneOverridesRaw;
        if (zoRaw && zoRaw.collisionEnabled === false) {
            mc.collisionEnabled = false;
            return;
        }

        mc.collisionEnabled = !!this.collisionEnabled;

        const z0 = this.zoomBoostStart;
        const z1 = Math.max(z0 + 0.001, this.zoomBoostFull);
        const k = clamp((dist - z0) / (z1 - z0), 0, 1);

        const camR = this.camRadius * (1.0 + this.zoomRadiusBoost * k);
        const nearR = this.nearRadius * (1.0 + this.zoomNearBoost * k);
        const floorPad = this.floorPadding * (1.0 + this.zoomFloorBoost * k);

        mc.camRadius = camR;
        mc.nearRadius = nearR;
        mc.pearK = this.pearK;
        mc.pearSamples = this.pearSamples;

        mc.surfacePadding = this.surfacePadding;
        mc.obstaclePasses = this.obstaclePasses;

        mc.useTerrainHeight = !!this.useTerrainHeight;
        mc.terrainWorld = !!this.terrainWorld;

        mc.floorPadding = floorPad;
        mc.slopePadScale = this.slopePadScale;

        mc.groundRayLift = this.groundRayLift;
        mc.maxRayLenDown = this.maxRayLenDown;
        mc.groundSnapPen = this.groundSnapPen;

        mc.debugCapsule = !!this.debugCapsule;
        mc.debugGroundCapsule = !!this.debugGroundCapsule;
    }

    _readMove(ctx) {
        const mv = ctx.moveDir || ctx.move || ctx.motion || null;
        const vel = ctx.bodyVel || ctx.vel || null;

        let dx = 0, dz = 0, sp = 0;

        if (mv) {
            dx = +U.vx(mv, 0) || 0;
            dz = +U.vz(mv, 0) || 0;
            const len = Math.hypot(dx, dz);
            if (len > 1e-6) {
                dx /= len;
                dz /= len;
            }
            sp = clamp(+mv.speed || +mv.mag || +mv.length || 0, 0, 1);
        } else if (vel) {
            const vx = +U.vx(vel, 0) || 0;
            const vz = +U.vz(vel, 0) || 0;
            const len = Math.hypot(vx, vz);
            if (len > 1e-6) {
                dx = vx / len;
                dz = vz / len;
            }
            sp = clamp(len / 6.0, 0, 1);
        }

        return {dx, dz, sp};
    }

    _softClampPitch(p) {
        const lo = this.pitchMin, hi = this.pitchMax;
        p = clamp(p, lo - 0.35, hi + 0.35);

        if (p < lo) {
            const t = clamp((lo - p) / Math.max(1e-6, this.pitchSoft), 0, 1);
            p = lo - (1.0 - smoothstep(1.0 - t)) * this.pitchSoft;
        } else if (p > hi) {
            const t = clamp((p - hi) / Math.max(1e-6, this.pitchSoft), 0, 1);
            p = hi + (1.0 - smoothstep(1.0 - t)) * this.pitchSoft;
        }
        return clamp(p, lo, hi);
    }

    _updateShoulderSide(ctx) {
        const inp = ctx.input || null;
        const swap = !!(ctx.shoulderSwap || (inp && inp.shoulderSwap));
        if (swap && !this._swapHeld) {
            this._shoulderSide = -this._shoulderSide;
            this._swapHeld = true;
        } else if (!swap) {
            this._swapHeld = false;
        }

        const zo = ctx.zoneOverrides;
        if (zo && (zo.shoulderSide === -1 || zo.shoulderSide === 1)) {
            this._shoulderSide = zo.shoulderSide;
        }
    }

    update(ctx) {
        const dt = Math.max(0, +ctx.dt || 0);

        const p = ctx.bodyPos;
        // Pre-mode zone-only overrides are exposed as ctx.zoneOverrides.
        const zo = ctx.zoneOverrides;

        const isAiming = !!(ctx.aiming || (ctx.input && ctx.input.aiming));

        const po = (zo && zo.pivotOffset) ? zo.pivotOffset : this.pivotOffset;

        const baseShoulder = (zo && zo.shoulderX != null) ? +zo.shoulderX : this.shoulderX;
        const aimShoulder = (zo && zo.shoulderAimX != null) ? +zo.shoulderAimX : this.shoulderAimX;
        const lift = (zo && zo.verticalLift != null) ? +zo.verticalLift : this.verticalLift;

        this._updateShoulderSide(ctx);

        let yaw = +ctx.look.yaw || 0;
        let pitch = +ctx.look.pitch || 0;

        pitch = this._softClampPitch(pitch);

        if (this.recenterEnabled) {
            const mv = this._readMove(ctx);
            if (mv.sp > 0.08) {
                const moveYaw = Math.atan2(mv.dx, mv.dz);
                let dy = yaw - moveYaw;
                while (dy > Math.PI) dy -= Math.PI * 2;
                while (dy < -Math.PI) dy += Math.PI * 2;

                const abs = Math.abs(dy);
                if (abs > this.recenterDeadZone) {
                    const over = clamp(abs - this.recenterDeadZone, 0, this.recenterMax);
                    const pull = smoothstep(over / Math.max(1e-6, this.recenterMax));
                    const a = expAlpha(this.recenterRate * (0.35 + 0.65 * mv.sp), dt) * pull;
                    yaw = lerp(yaw, moveYaw, a);
                }
            }
        }

        const sinY = Math.sin(yaw);
        const cosY = Math.cos(yaw);

        const rx = cosY, rz = -sinY;
        const fx = sinY, fz = cosY;

        const basePx = U.vx(p, 0) + po.x;
        const basePy = U.vy(p, 0) + po.y;
        const basePz = U.vz(p, 0) + po.z;

        const mv = this._readMove(ctx);
        const leadMax = isAiming ? this.forwardLeadAim : this.forwardLead;
        const leadK = leadMax * mv.sp;

        const leadAx = expAlpha(this.leadSmooth, dt);
        const desiredLeadX = (mv.sp > 1e-4) ? mv.dx * leadK : fx * (leadK * 0.35);
        const desiredLeadZ = (mv.sp > 1e-4) ? mv.dz * leadK : fz * (leadK * 0.35);

        if (!this._init) {
            this._lead.x = desiredLeadX;
            this._lead.z = desiredLeadZ;
        } else {
            this._lead.x = lerp(this._lead.x, desiredLeadX, leadAx);
            this._lead.z = lerp(this._lead.z, desiredLeadZ, leadAx);
        }

        const aimBlend = isAiming ? 1.0 : 0.0;
        const shoulder = lerp(baseShoulder, aimShoulder, aimBlend) * this._shoulderSide;

        const rawPx = basePx + rx * shoulder + this._lead.x;
        const rawPy = basePy;
        const rawPz = basePz + rz * shoulder + this._lead.z;

        if (!this._init) {
            this._init = true;
            this._pivot.x = rawPx;
            this._pivot.y = rawPy;
            this._pivot.z = rawPz;

            this._cam.x = rawPx;
            this._cam.y = rawPy + lift;
            this._cam.z = rawPz;
        } else {
            const ax = expAlpha(this.pivotSmoothPos, dt);
            const ay = expAlpha(this.pivotSmoothY, dt);

            this._pivot.x = lerp(this._pivot.x, rawPx, ax);
            this._pivot.y = lerp(this._pivot.y, rawPy, ay);
            this._pivot.z = lerp(this._pivot.z, rawPz, ax);
        }

        ctx.target.x = this._pivot.x;
        ctx.target.y = this._pivot.y;
        ctx.target.z = this._pivot.z;

        const dist = Math.max(0.05, +ctx.zoom.value());
        this._applyCollisionOverrides(ctx, dist);

        const p2 = pitch * this.pitchOrbitScale;
        const cp = Math.cos(p2);
        const sp = Math.sin(p2);

        const horiz = dist * cp;

        const outX = this._pivot.x - sinY * horiz;
        const outZ = this._pivot.z - cosY * horiz;
        const outY = this._pivot.y + lift + sp * dist;

        const cx = expAlpha(this.camSmoothPos, dt);
        const cy = expAlpha(this.camSmoothY, dt);

        this._cam.x = lerp(this._cam.x, outX, cx);
        this._cam.y = lerp(this._cam.y, outY, cy);
        this._cam.z = lerp(this._cam.z, outZ, cx);

        ctx.outPos.x = this._cam.x;
        ctx.outPos.y = this._cam.y;
        ctx.outPos.z = this._cam.z;

        if (ctx.outLook) {
            ctx.outLook.yaw = yaw;
            ctx.outLook.pitch = pitch;
        }
    }
}

module.exports = ThirdPersonCameraMode;

File: kalitech-engine/src/main/resources/kalitech/engine/modules/Camera/modes/third.lua
local M = {}
local luaRuntime = require("@builtin/lua_runtime")
local LuaClass = luaRuntime.LuaClass
local LuaNumber = luaRuntime.LuaNumber
U = require("../camUtil.lua")
function clamp(self, v, lo, hi)
    local lua_temp_1
    if v < lo then
        lua_temp_1 = lo
    else
        local lua_temp_0
        if v > hi then
            lua_temp_0 = hi
        else
            lua_temp_0 = v
        end
        lua_temp_1 = lua_temp_0
    end
    return lua_temp_1
end
function lerp(self, a, b, t)
    return a + (b - a) * t
end
function smoothstep(self, t)
    t = clamp(_G, t, 0, 1)
    return t * t * (3 - 2 * t)
end
function expAlpha(self, rate, dt)
    return 1 - math.exp(-math.max(0, rate) * math.max(0, dt))
end
ThirdPersonCameraMode = LuaClass()
ThirdPersonCameraMode.name = "ThirdPersonCameraMode"
function ThirdPersonCameraMode.prototype.lua_constructor(self)
    self.id = "third"
    self.meta = {supportsZoom = true, hasCollision = true, numRays = 8, playerModelVisible = true}
    self.pivotOffset = {x = 0, y = 1.46, z = 0}
    self.shoulderX = 0.42
    self.shoulderAimX = 0.26
    self.verticalLift = 0.14
    self.forwardLead = 0.28
    self.forwardLeadAim = 0.08
    self.leadSmooth = 12
    self.pivotSmoothPos = 22
    self.pivotSmoothY = 26
    self.camSmoothPos = 18
    self.camSmoothY = 20
    self.recenterEnabled = true
    self.recenterRate = 2.2
    self.recenterDeadZone = 0.38
    self.recenterMax = 1.25
    self.pitchOrbitScale = 1
    self.pitchMin = -1.15
    self.pitchMax = 0.9
    self.pitchSoft = 0.18
    self.collisionEnabled = true
    self.camRadius = 0.28
    self.nearRadius = 0.06
    self.pearK = 1.9
    self.pearSamples = 9
    self.surfacePadding = 0.08
    self.obstaclePasses = 2
    self.useTerrainHeight = true
    self.terrainWorld = true
    self.floorPadding = 0.22
    self.slopePadScale = 0.45
    self.groundRayLift = 1.25
    self.maxRayLenDown = 12
    self.groundSnapPen = 0.6
    self.zoomRadiusBoost = 0.1
    self.zoomNearBoost = 0.06
    self.zoomFloorBoost = 0.2
    self.zoomBoostStart = 6
    self.zoomBoostFull = 26
    self.debugCapsule = true
    self.debugGroundCapsule = true
    self._init = false
    self._pivot = {x = 0, y = 0, z = 0}
    self._cam = {x = 0, y = 0, z = 0}
    self._lead = {x = 0, y = 0, z = 0}
    self._shoulderSide = 1
end
function ThirdPersonCameraMode.prototype._ensureModeConfig(self, ctx)
    local mc = ctx.modeConfig
    if not mc then
        mc = {}
        ctx.modeConfig = mc
    end
    return mc
end
function ThirdPersonCameraMode.prototype._applyCollisionOverrides(self, ctx, dist)
    local mc = self:_ensureModeConfig(ctx)
    local zoRaw = ctx.zoneOverridesRaw
    if zoRaw and zoRaw.collisionEnabled == false then
        mc.collisionEnabled = false
        return
    end
    mc.collisionEnabled = not not self.collisionEnabled
    local z0 = self.zoomBoostStart
    local z1 = math.max(z0 + 0.001, self.zoomBoostFull)
    local k = clamp(_G, (dist - z0) / (z1 - z0), 0, 1)
    local camR = self.camRadius * (1 + self.zoomRadiusBoost * k)
    local nearR = self.nearRadius * (1 + self.zoomNearBoost * k)
    local floorPad = self.floorPadding * (1 + self.zoomFloorBoost * k)
    mc.camRadius = camR
    mc.nearRadius = nearR
    mc.pearK = self.pearK
    mc.pearSamples = self.pearSamples
    mc.surfacePadding = self.surfacePadding
    mc.obstaclePasses = self.obstaclePasses
    mc.useTerrainHeight = not not self.useTerrainHeight
    mc.terrainWorld = not not self.terrainWorld
    mc.floorPadding = floorPad
    mc.slopePadScale = self.slopePadScale
    mc.groundRayLift = self.groundRayLift
    mc.maxRayLenDown = self.maxRayLenDown
    mc.groundSnapPen = self.groundSnapPen
    mc.debugCapsule = not not self.debugCapsule
    mc.debugGroundCapsule = not not self.debugGroundCapsule
end
function ThirdPersonCameraMode.prototype._readMove(self, ctx)
    local mv = ctx.moveDir or ctx.move or ctx.motion or nil
    local vel = ctx.bodyVel or ctx.vel or nil
    local dx = 0
    local dz = 0
    local sp = 0
    if mv then
        dx = LuaNumber(U:vx(mv, 0)) or 0
        dz = LuaNumber(U:vz(mv, 0)) or 0
        local len = KMath:hypot(dx, dz)
        if len > 0.000001 then
            dx = dx / len
            dz = dz / len
        end
        sp = clamp(
            _G,
            LuaNumber(mv.speed) or LuaNumber(mv.mag) or LuaNumber(KLength(mv)) or 0,
            0,
            1
        )
    elseif vel then
        local vx = LuaNumber(U:vx(vel, 0)) or 0
        local vz = LuaNumber(U:vz(vel, 0)) or 0
        local len = KMath:hypot(vx, vz)
        if len > 0.000001 then
            dx = vx / len
            dz = vz / len
        end
        sp = clamp(_G, len / 6, 0, 1)
    end
    return {dx = dx, dz = dz, sp = sp}
end
function ThirdPersonCameraMode.prototype._softClampPitch(self, p)
    local lo = self.pitchMin
    local hi = self.pitchMax
    p = clamp(_G, p, lo - 0.35, hi + 0.35)
    if p < lo then
        local t = clamp(
            _G,
            (lo - p) / math.max(0.000001, self.pitchSoft),
            0,
            1
        )
        p = lo - (1 - smoothstep(_G, 1 - t)) * self.pitchSoft
    elseif p > hi then
        local t = clamp(
            _G,
            (p - hi) / math.max(0.000001, self.pitchSoft),
            0,
            1
        )
        p = hi + (1 - smoothstep(_G, 1 - t)) * self.pitchSoft
    end
    return clamp(_G, p, lo, hi)
end
function ThirdPersonCameraMode.prototype._updateShoulderSide(self, ctx)
    local inp = ctx.input or nil
    local swap = not not (ctx.shoulderSwap or inp and inp.shoulderSwap)
    if swap and not self._swapHeld then
        self._shoulderSide = -self._shoulderSide
        self._swapHeld = true
    elseif not swap then
        self._swapHeld = false
    end
    local zo = ctx.zoneOverrides
    if zo and (zo.shoulderSide == -1 or zo.shoulderSide == 1) then
        self._shoulderSide = zo.shoulderSide
    end
end
function ThirdPersonCameraMode.prototype.update(self, ctx)
    local dt = math.max(
        0,
        LuaNumber(ctx.dt) or 0
    )
    local p = ctx.bodyPos
    local zo = ctx.zoneOverrides
    local isAiming = not not (ctx.aiming or ctx.input and ctx.input.aiming)
    local lua_temp_2
    if zo and zo.pivotOffset then
        lua_temp_2 = zo.pivotOffset
    else
        lua_temp_2 = self.pivotOffset
    end
    local po = lua_temp_2
    local lua_temp_3
    if zo and zo.shoulderX ~= nil then
        lua_temp_3 = LuaNumber(zo.shoulderX)
    else
        lua_temp_3 = self.shoulderX
    end
    local baseShoulder = lua_temp_3
    local lua_temp_4
    if zo and zo.shoulderAimX ~= nil then
        lua_temp_4 = LuaNumber(zo.shoulderAimX)
    else
        lua_temp_4 = self.shoulderAimX
    end
    local aimShoulder = lua_temp_4
    local lua_temp_5
    if zo and zo.verticalLift ~= nil then
        lua_temp_5 = LuaNumber(zo.verticalLift)
    else
        lua_temp_5 = self.verticalLift
    end
    local lift = lua_temp_5
    self:_updateShoulderSide(ctx)
    local yaw = LuaNumber(ctx.look.yaw) or 0
    local pitch = LuaNumber(ctx.look.pitch) or 0
    pitch = self:_softClampPitch(pitch)
    if self.recenterEnabled then
        local mv = self:_readMove(ctx)
        if mv.sp > 0.08 then
            local moveYaw = math.atan2(mv.dx, mv.dz)
            local dy = yaw - moveYaw
            while dy > math.pi do
                dy = dy - math.pi * 2
            end
            while dy < -math.pi do
                dy = dy + math.pi * 2
            end
            local abs = math.abs(dy)
            if abs > self.recenterDeadZone then
                local over = clamp(_G, abs - self.recenterDeadZone, 0, self.recenterMax)
                local pull = smoothstep(
                    _G,
                    over / math.max(0.000001, self.recenterMax)
                )
                local a = expAlpha(_G, self.recenterRate * (0.35 + 0.65 * mv.sp), dt) * pull
                yaw = lerp(_G, yaw, moveYaw, a)
            end
        end
    end
    local sinY = math.sin(yaw)
    local cosY = math.cos(yaw)
    local rx = cosY
    local rz = -sinY
    local fx = sinY
    local fz = cosY
    local basePx = U:vx(p, 0) + po.x
    local basePy = U:vy(p, 0) + po.y
    local basePz = U:vz(p, 0) + po.z
    local mv = self:_readMove(ctx)
    local lua_isAiming_6
    if isAiming then
        lua_isAiming_6 = self.forwardLeadAim
    else
        lua_isAiming_6 = self.forwardLead
    end
    local leadMax = lua_isAiming_6
    local leadK = leadMax * mv.sp
    local leadAx = expAlpha(_G, self.leadSmooth, dt)
    local lua_temp_7
    if mv.sp > 0.0001 then
        lua_temp_7 = mv.dx * leadK
    else
        lua_temp_7 = fx * (leadK * 0.35)
    end
    local desiredLeadX = lua_temp_7
    local lua_temp_8
    if mv.sp > 0.0001 then
        lua_temp_8 = mv.dz * leadK
    else
        lua_temp_8 = fz * (leadK * 0.35)
    end
    local desiredLeadZ = lua_temp_8
    if not self._init then
        self._lead.x = desiredLeadX
        self._lead.z = desiredLeadZ
    else
        self._lead.x = lerp(_G, self._lead.x, desiredLeadX, leadAx)
        self._lead.z = lerp(_G, self._lead.z, desiredLeadZ, leadAx)
    end
    local aimBlend = isAiming and 1 or 0
    local shoulder = lerp(_G, baseShoulder, aimShoulder, aimBlend) * self._shoulderSide
    local rawPx = basePx + rx * shoulder + self._lead.x
    local rawPy = basePy
    local rawPz = basePz + rz * shoulder + self._lead.z
    if not self._init then
        self._init = true
        self._pivot.x = rawPx
        self._pivot.y = rawPy
        self._pivot.z = rawPz
        self._cam.x = rawPx
        self._cam.y = rawPy + lift
        self._cam.z = rawPz
    else
        local ax = expAlpha(_G, self.pivotSmoothPos, dt)
        local ay = expAlpha(_G, self.pivotSmoothY, dt)
        self._pivot.x = lerp(_G, self._pivot.x, rawPx, ax)
        self._pivot.y = lerp(_G, self._pivot.y, rawPy, ay)
        self._pivot.z = lerp(_G, self._pivot.z, rawPz, ax)
    end
    ctx.target.x = self._pivot.x
    ctx.target.y = self._pivot.y
    ctx.target.z = self._pivot.z
    local dist = math.max(
        0.05,
        LuaNumber(ctx.zoom:value())
    )
    self:_applyCollisionOverrides(ctx, dist)
    local p2 = pitch * self.pitchOrbitScale
    local cp = math.cos(p2)
    local sp = math.sin(p2)
    local horiz = dist * cp
    local outX = self._pivot.x - sinY * horiz
    local outZ = self._pivot.z - cosY * horiz
    local outY = self._pivot.y + lift + sp * dist
    local cx = expAlpha(_G, self.camSmoothPos, dt)
    local cy = expAlpha(_G, self.camSmoothY, dt)
    self._cam.x = lerp(_G, self._cam.x, outX, cx)
    self._cam.y = lerp(_G, self._cam.y, outY, cy)
    self._cam.z = lerp(_G, self._cam.z, outZ, cx)
    ctx.outPos.x = self._cam.x
    ctx.outPos.y = self._cam.y
    ctx.outPos.z = self._cam.z
    if ctx.outLook then
        ctx.outLook.yaw = yaw
        ctx.outLook.pitch = pitch
    end
end
M = ThirdPersonCameraMode

return M
